editor/src/components/assets.ts

```diff
@@ -197,6 +210,181 @@ export function walkContentsTree(
+  tree: ProjectContentTreeRoot,
+  path: string,
+  file: ProjectFile,
+): ProjectContentTreeRoot {
+  const pathElements = getProjectContentKeyPathElements(path)
+
+  function createProjectContent(): ProjectContentsTree {
+    if (isDirectory(file)) {
+      return projectContentDirectory(path, file, {})
+    } else {
+      return projectContentFile(path, file)
+    }
+  }
+
+  function addAtCurrentIndex(
+    workingTreeRoot: ProjectContentTreeRoot,
+    index: number,
+  ): ProjectContentTreeRoot {
+    const isLastElement = index === pathElements.length - 1
+    const pathPart = pathElements[index]
+    const treePart = workingTreeRoot[pathPart]
+    if (treePart == null) {
+      if (isLastElement) {
+        return {
+          ...workingTreeRoot,
+          [pathPart]: createProjectContent(),
+        }
+      } else {
+        return {
+          ...workingTreeRoot,
+          [pathPart]: projectContentDirectory(
+            pathElementsToProjectContentKey(pathElements, index),
+            directory(),
+            addAtCurrentIndex({}, index + 1),
+          ),
+        }
+      }
+    } else {
+      if (isLastElement) {
+        return {
+          ...workingTreeRoot,
+          [pathPart]: createProjectContent(),
+        }
+      } else {
+        if (treePart.type === 'PROJECT_CONTENT_FILE') {
+          throw new Error(
+            `Attempting to add something at ${path} which is below a pre-existing file at ${treePart.fullPath}`,
+          )
+        } else {
+          return {
+            ...workingTreeRoot,
+            [pathPart]: projectContentDirectory(
+              treePart.fullPath,
+              treePart.directory,
+              addAtCurrentIndex(treePart.children, index + 1),
+            ),
+          }
+        }
+      }
+    }
+  }
+
+  return addAtCurrentIndex(tree, 0)
+}
+
+export function removeFromProjectContents(
+  projectContents: ProjectContentTreeRoot,
+  path: string,
+): ProjectContentTreeRoot {
+  const pathElements = getProjectContentKeyPathElements(path)
+
+  function removeWithIndex(
+    workingTreeRoot: ProjectContentTreeRoot,
+    pathIndex: number,
+  ): 'NO_CHANGE' | ProjectContentTreeRoot {
+    const isLastElement = pathIndex === pathElements.length - 1
+    const pathPart = pathElements[pathIndex]
+    const treePart = workingTreeRoot[pathPart]
+
+    if (treePart == null) {
+      // Reached a part of the tree that doesn't exist, therefore the target
+      // doesn't exist.
+      return 'NO_CHANGE'
+    } else {
+      if (isLastElement) {
+        let updatedWorkingTreeRoot: ProjectContentTreeRoot = {
+          ...workingTreeRoot,
+        }
+        delete updatedWorkingTreeRoot[pathPart]
+        return updatedWorkingTreeRoot
+      } else {
+        if (treePart.type === 'PROJECT_CONTENT_FILE') {
+          // Attempting to delete something which is hierarchically below a file.
+          // Which means it can't possibly exist.
+          return 'NO_CHANGE'
```

Comment:
shouldn't this throw an error?

Reply:
I don't believe so, it's just trying to delete a path that doesn't exist.